### PR TITLE
Check for function bodies in assisted functions

### PR DESCRIPTION
**Pull request by ZacSweers**
Ref #415

### CODE REVIEW

File: compiler/src/main/java/com/squareup/anvil/compiler/codegen/dagger/AssistedFactoryGenerator.kt [hunk 1]
```diff
@@ -440,11 +441,16 @@ internal class AssistedFactoryGenerator : PrivateCodeGenerator() {
...
-          "method but found multiple.",
-        element = this
-      )
+      else -> {
+        val foundFunctions = assistedFunctions.joinToString { func ->
+          "${func.name}(${func.parameterPairs.value.map { it.first }})"
+        }
+        throw AnvilCompilationException(
+          "The @AssistedFactory-annotated type should contain a single abstract, non-default " +
+            "method but found multiple: [$foundFunctions]",
```
Comment (ZacSweers) [inline compiler/src/main/java/com/squareup/anvil/compiler/codegen/dagger/AssistedFactoryGenerator.kt:450]: this matches dagger's error message as well

File: compiler/src/test/java/com/squareup/anvil/compiler/dagger/AssistedFactoryGeneratorTest.kt [hunk 2]
```diff
@@ -1615,6 +1616,44 @@ public final class AssistedServiceFactory_Impl implements AssistedServiceFactory
...
+  @Test fun `default functions do not count against SAM requirement`() {
+    prepareCompilation()
+      .apply {
+        kotlinCompilation.apply {
+          // Necessary so Dagger-compiler recognizes default functions too
+          @Suppress("SuspiciousCollectionReassignment")
+          kotlincArguments += "-Xjvm-default=all"
+        }
+      }
+      .configureAnvil()
```
Comment (vRallev) [inline compiler/src/test/java/com/squareup/anvil/compiler/dagger/AssistedFactoryGeneratorTest.kt:1628]: This is the problem.

File: compiler/src/test/java/com/squareup/anvil/compiler/dagger/AssistedFactoryGeneratorTest.kt [hunk 3]
```diff
@@ -1615,6 +1616,44 @@ public final class AssistedServiceFactory_Impl implements AssistedServiceFactory
...
+            return get()
+          }
+          
+          fun create(string: String): AssistedService {
+            return create()
+          }
+        }
+        """
+      ) {
+        assertThat(exitCode).isEqualTo(ExitCode.OK)
```
Comment (vRallev) [inline compiler/src/test/java/com/squareup/anvil/compiler/dagger/AssistedFactoryGeneratorTest.kt:1653]: Can you import `OK` here directly?

File: compiler/src/test/java/com/squareup/anvil/compiler/dagger/AssistedFactoryGeneratorTest.kt [hunk 4]
```diff
@@ -1615,6 +1616,44 @@ public final class AssistedServiceFactory_Impl implements AssistedServiceFactory
     }
   }
 
+  @Test fun `default functions do not count against SAM requirement`() {
+    prepareCompilation()
+      .apply {
+        kotlinCompilation.apply {
```
Comment (vRallev) [inline compiler/src/test/java/com/squareup/anvil/compiler/dagger/AssistedFactoryGeneratorTest.kt:1622]: The 2nd `apply` isn't necessary here. You can remove it.

File: compiler/src/test/java/com/squareup/anvil/compiler/dagger/AssistedFactoryGeneratorTest.kt (hunk 2 above)
Comment (ZacSweers) [reply compiler/src/test/java/com/squareup/anvil/compiler/dagger/AssistedFactoryGeneratorTest.kt:1628]:
```suggestion
```

**Review by vRallev — APPROVED** [review]

**Pull request merged**